### DISCUSSION TITLE
[client-preset] Reason for config restrictions on `client-preset`

### DISCUSSION THREAD

**Discussion by amaretinus** [dotansimha/graphql-code-generator#8993] [Q&A]
Found this code in the `@graphql-codegen` build.
The docs specified the configs supported for `client-preset`, I just wanted to make sure.

```
// the `client` preset is restricting the config options inherited from `typescript`, `typescript-operations` and others.
const forwardedConfig = {
  scalars: options.config.scalars,
  defaultScalarType: options.config.defaultScalarType,
  strictScalars: options.config.strictScalars,
  namingConvention: options.config.namingConvention,
  useTypeImports: options.config.useTypeImports,
  skipTypename: options.config.skipTypename,
  arrayInputCoercion: options.config.arrayInputCoercion,
  enumsAsTypes: options.config.enumsAsTypes,
  dedupeFragments: options.config.dedupeFragments,
  nonOptionalTypename: options.config.nonOptionalTypename,
  avoidOptionals: options.config.avoidOptionals,
};
```

There are a lot of useful configs that I need to use in my current project, and these doesn't suffice.
May I know the rationale behind the restriction of the other config options?

**Comment by n1ru4l** [comment]
> I need to use in my current project
Can you elaborate on why each of these configuration options is crucial for your project? Why restricted the number of options intentionally as we did not see any benefit in having them and we want to remove a lot of these in a future rewrite of the internals.

We are happy to re-introduce specific configuration options given there is a use-case.

**Reply by ArnaudD** [reply]
Hi @n1ru4l,
In my project `enumValues` is really a must have to match graphql enums to existing TS (string) enums.
I made a patch to my local package (using `yarn patch`) and it seems to works well. Will it cause me unexpected troubles ? Could it be re-introduced in the client-preset options ? I can make a PR if you like.

**Reply by jthoward64** [reply]
I don't understand why there isn't an option to pass through arbitrary config options, say something like `configOverride` or similar. There's a lot of configuration options in `typescript` and `typescript-operations` that are really helpful that are simply unavailable and will hinder migration. For now I'm just gonna have to go the `yarn patch` route.

**Reply by KBorders01** [reply]
I also need to use enumValues because I'm refactoring a bunch of code that has references to existing enum declarations.

**Reply by Stephen2** [reply]
Hi, there are PRs https://github.com/dotansimha/graphql-code-generator/pull/9574 and issues https://github.com/dotansimha/graphql-code-generator/issues/8562 modifying the available options, that aren't getting much attention

**Comment by Akkuma** [comment]
`client-preset` prevents useful community plugins like `typescript-enum-array` from being configured at all. It seems wrong to prevent passing config options to these plugins.

**Comment by jthoward64** [comment]
FWIW I've been using this yarn patch and it has worked well for me:

> @graphql-codegen-client-preset-npm-4.1.0-ffbe3e3006.patch
```patch
diff --git a/cjs/index.js b/cjs/index.js
index 8424b6a11586092665872ab1e190b57602efe44b..9e6f342e1fe583a40863b029db887fdd95c1f360 100644
--- a/cjs/index.js
+++ b/cjs/index.js
@@ -182,7 +182,7 @@ exports.preset = {
                 schema: options.schema,
                 config: {
                     inlineFragmentTypes: isMaskingFragments ? 'mask' : options.config['inlineFragmentTypes'],
-                    ...forwardedConfig,
+                    ...options.config,
                 },
                 documents: sources,
                 documentTransforms: options.documentTransforms,
diff --git a/esm/index.js b/esm/index.js
index be26a67a1b827896d16813b22b9a45aec5ac5586..c7cb26b337ef39b4270aff021cfe9ccb148e0848 100644
--- a/esm/index.js
+++ b/esm/index.js
@@ -177,7 +177,7 @@ export const preset = {
                 schema: options.schema,
                 config: {
                     inlineFragmentTypes: isMaskingFragments ? 'mask' : options.config['inlineFragmentTypes'],
-                    ...forwardedConfig,
+                    ...options.config,
                 },
                 documents: sources,
                 documentTransforms: options.documentTransforms,
```

Just be careful not to set anything that breaks the preset

**Reply by Akkuma** [reply]
This worked like a charm with the same approach in `pnpm`.

**Comment by 2coo** [comment]
They recommend `client-preset` but they resticted `enumAsConst` option in that.
¯\\\_(ツ)_/¯

**Comment by jthoward64** [comment]
For what it's work I have been using the following configs that are not exposed by Client Preset for almost a year now without issue:

```ts
  enumsAsConst: true,
  immutableTypes: true,
  enumValues: {
    ...
  },
```

enumValues is especially useful to use my common library's exported enums rather than duplicating them
